### PR TITLE
fix(core): prevent memory keys from shadowing reserved context in edgee evaluation

### LINKED ISSUE #595: [Bug]: Missing Validation in Edge Condition Evaluation
### Location
- **File**: `core/framework/graph/edge.py`
- **Method**: `_evaluate_condition()`
- **Line**: ~175-177

### Description
The `safe_eval()` function is called with a context dictionary that includes unpacked memory keys. If memory contains keys that conflict with built-in names or context variables (e.g., `output`, `memory`, `result`), the evaluation might produce unexpected results or errors.

### Code Snippet
```python
context = {
    "output": output,
    "memory": memory,
    "result": output.get("result"),
    "true": True,
    "false": False,
    **memory,  # ⚠️ Could overwrite context variables
}
```

### Impact
- **Logic Error**: Memory keys could shadow context variables
- **Production Risk**: Low - Depends on memory key naming
- **User Experience**: Unexpected behavior in edge condition evaluation

### Suggested Fix
```python
# Build evaluation context with conflict detection
base_context = {
    "output": output,
    "memory": memory,
    "result": output.get("result"),
    "true": True,
    "false": False,
}

# Check for conflicts before unpacking memory
conflicts = set(base_context.keys()) & set(memory.keys())
if conflicts:
    logger.warning(f"Memory keys conflict with context variables: {conflicts}")

context = {**base_context, **memory}
```

---

### PR DESCRIPTION
**Pull request by ANKIT0017**
Resolves Issue #595. Filters memory keys to protect 'output', 'result', and 'memory' variables during condition evaluation. Adds full regression test suite.

## Description

This PR resolves a bug in the edge condition evaluation logic where agent memory keys could unknowingly overwrite critical framework variables like output, result, and memory. By implementing a reserved keys filter, the framework now protects these context variables while still allowing developers to access shadowed memory values explicitly via the memory dictionary.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)

## Related Issues

Fixes #595

## Changes Made

-Protected reserved framework variables (output, result, memory, true, false) during context construction
-Filtered out conflicting keys from the agent's memory before unpacking into evaluation context
-Added warning logs when memory keys conflict with reserved variables
-Added support for lowercase true and false in conditional expressions
-Updated CHANGELOG.md to document the fix

## Testing

-Successfully ran target unit tests: pytest core/tests/test_edge_conditions.py
-Verified protection against variable shadowing
-Verified explicit access via memory dictionary
-Verified lowercase boolean evaluation

## Checklist

-  My code follows the project's style guidelines
-  I have performed a self-review of my code
-  I have commented my code, particularly in hard-to-understand areas
-  My changes generate no new warnings
-  I have added tests that prove my fix is effective or that my feature works
-  New and existing unit tests pass locally with my changes

### CODE REVIEW

File: core/framework/graph/edge.py [hunk 1]
```diff
@@ -161,17 +161,34 @@ def _evaluate_condition(
...
-            **memory,  # Unpack memory keys directly into context
         }
 
+        # Filter out reserved keys from memory to prevent shadowing
+        safe_memory = {k: v for k, v in memory.items() if k not in RESERVED_KEYS}
+
+        # Warn about conflicts
+        conflicts = set(memory.keys()) - set(safe_memory.keys())
+        if conflicts:
+            import logging
```
Comment (Hundao) [inline core/framework/graph/edge.py]: Move import logging to the top of the file — shouldn't be inside a conditional block.

File: core/framework/graph/edge.py [hunk 2]
```diff
@@ -161,17 +161,34 @@ def _evaluate_condition(
...
+            logger = logging.getLogger(__name__)
+            logger.warning(
+                f"      ⚠ Memory keys {conflicts} conflict with reserved context variables. "
```
Comment (Hundao) [inline core/framework/graph/edge.py]: Remove the leading whitespace in the warning message ("      ⚠ Memory keys...").

File: core/framework/graph/edge.py [hunk 3]
```diff
@@ -161,17 +161,34 @@ def _evaluate_condition(
...
             "true": True,   # Allow lowercase true/false in conditions
             "false": False,
-            **memory,  # Unpack memory keys directly into context
         }
 
+        # Filter out reserved keys from memory to prevent shadowing
+        safe_memory = {k: v for k, v in memory.items() if k not in RESERVED_KEYS}
+
+        # Warn about conflicts
+        conflicts = set(memory.keys()) - set(safe_memory.keys())
```
Comment (Hundao) [inline core/framework/graph/edge.py]: Unnecessary indirection — conflicts can be simplified to set(memory.keys()) & RESERVED_KEYS.

**Review by Hundao — COMMENTED** [review]
Hi @ANKIT0017,

Please have a look and solve the conflict, thanks!
